src/iceberg/schema_util.cc

```diff
@@ -0,0 +1,213 @@
+  }
+  for (auto& child_projection : field_projection.children) {
+    if (child_projection.kind == FieldProjection::Kind::kProjected) {
+      child_projection.from =
+          local_index_to_pruned_index.at(std::get<1>(child_projection.from));
+    }
+  }
+}
+
+Result<FieldProjection> ProjectNested(const Type& expected_type, const Type& source_type,
+                                      bool prune_source) {
+  if (!expected_type.is_nested()) {
+    return InvalidSchema("Expected a nested type, but got {}", expected_type);
+  }
+  if (expected_type.type_id() != source_type.type_id()) {
+    return InvalidSchema("Expected {}, but got {}", expected_type, source_type);
+  }
+
+  const auto& expected_fields =
+      internal::checked_cast<const NestedType&>(expected_type).fields();
+  const auto& source_fields =
+      internal::checked_cast<const NestedType&>(source_type).fields();
+
+  // Build a map from field id to source field info including its local offset in
+  // the current nesting level.
+  struct SourceFieldInfo {
+    size_t local_index;
+    const SchemaField* field;
+  };
+  std::unordered_map<int32_t, SourceFieldInfo> source_field_map;
+  source_field_map.reserve(source_fields.size());
+  for (size_t i = 0; i < source_fields.size(); ++i) {
+    const auto& field = source_fields[i];
+    if (const auto [iter, inserted] = source_field_map.emplace(
+            std::piecewise_construct, std::forward_as_tuple(field.field_id()),
+            std::forward_as_tuple(i, &field));
+        !inserted) [[unlikely]] {
+      return InvalidSchema("Duplicate field id found, prev: {}, curr: {}",
+                           *iter->second.field, field);
+    }
+  }
+
+  FieldProjection result;
+  result.children.reserve(expected_fields.size());
+
+  for (const auto& expected_field : expected_fields) {
+    int32_t field_id = expected_field.field_id();
+    FieldProjection child_projection;
+
+    if (auto iter = source_field_map.find(field_id); iter != source_field_map.cend()) {
+      if (expected_field.type()->is_nested()) {
+        ICEBERG_ASSIGN_OR_RAISE(auto nested_projection,
+                                ProjectNested(*expected_field.type(),
+                                              *iter->second.field->type(), prune_source));
+        child_projection.children.emplace_back(std::move(nested_projection));
+      } else {
+        ICEBERG_RETURN_UNEXPECTED(
+            ValidateSchemaEvolution(*expected_field.type(), *iter->second.field->type()));
+      }
+      // If `prune_source` is false, all fields will be read so the local index
+      // is exactly the position to read data. Otherwise, the local index is computed
+      // by pruning all non-projected fields
+      child_projection.from = iter->second.local_index;
+      child_projection.kind = FieldProjection::Kind::kProjected;
+    } else if (MetadataColumns::IsMetadataColumn(field_id)) {
+      child_projection.kind = FieldProjection::Kind::kMetadata;
```

Comment:
Should this be checked before `source_field_map.find(field_id)`?

Reply:
No, in most cases fields should be found from the source schema. Checking metadata column first may introduces more misses.